data/src/main/java/io/homeassistant/companion/android/data/url/UrlRepositoryImpl.kt

```diff
@@ -0,0 +1,97 @@
+package io.homeassistant.companion.android.data.url
+
+import io.homeassistant.companion.android.data.LocalStorage
+import io.homeassistant.companion.android.domain.MalformedHttpUrlException
+import io.homeassistant.companion.android.domain.url.UrlRepository
+import java.net.URL
+import javax.inject.Inject
+import javax.inject.Named
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
+
+class UrlRepositoryImpl @Inject constructor(
+    @Named("url") private val localStorage: LocalStorage
+) : UrlRepository {
+
+    companion object {
+        private const val PREF_CLOUD_URL = "cloud_url"
+        private const val PREF_REMOTE_URL = "remote_url"
+        private const val PREF_WEBHOOK_ID = "webhook_id"
+        private const val PREF_LOCAL_URL = "local_url"
+        private const val PREF_WIFI_SSID = "wifi_ssid"
+    }
+
+    override suspend fun getApiUrls(): Array<URL> {
+        val retVal = ArrayList<URL>()
+        val webhook = localStorage.getString(PREF_WEBHOOK_ID)
+
+        localStorage.getString(PREF_CLOUD_URL)?.let {
+            retVal.add(it.toHttpUrl().toUrl())
+        }
+
+        localStorage.getString(PREF_REMOTE_URL)?.let {
+            retVal.add(
+                it.toHttpUrl().newBuilder()
+                    .addPathSegments("api/webhook/$webhook")
+                    .build()
+                    .toUrl()
+            )
+        }
+
+        localStorage.getString(PREF_LOCAL_URL)?.let {
+            retVal.add(
+                it.toHttpUrl().newBuilder()
+                    .addPathSegments("api/webhook/$webhook")
+                    .build()
+                    .toUrl()
+            )
+        }
+
+        return retVal.toTypedArray()
+    }
+
+    override suspend fun saveRegistrationUrls(
+        cloudHookUrl: String?,
+        remoteUiUrl: String?,
+        webhookId: String
+    ) {
+        localStorage.putString(PREF_CLOUD_URL, cloudHookUrl)
+        localStorage.putString(PREF_REMOTE_URL, remoteUiUrl)
+        localStorage.putString(PREF_WEBHOOK_ID, webhookId)
+    }
+
+    override suspend fun getUrl(isInternal: Boolean): URL? {
+        val internal = localStorage.getString(PREF_LOCAL_URL)?.toHttpUrlOrNull()?.toUrl()
+        val external = localStorage.getString(PREF_REMOTE_URL)?.toHttpUrlOrNull()?.toUrl()
+        return if (isInternal) {
+            internal ?: external
+        } else {
+            external ?: internal
```

Comment:
If we're external, would it really help in ever returning the internal url ?

Reply:
It really depends on peoples setups, if someone manually sets up the url during onboarding that gets persisted as the internal address.  However, without the wifi ssid set they are technically external.  So in that case they need the internal address returned.